CMakeLists.txt

```diff
@@ -33,40 +47,29 @@ else()
   set(hunter_config "")
 endif()
 
-set(package_deps)
-
 if(CMAKE_CXX_COMPILER_ID STREQUAL "GNU" AND
     CMAKE_CXX_COMPILER_VERSION VERSION_LESS "4.9")
   set(boost_components regex)
   hunter_add_package(Boost COMPONENTS regex)
-  list(APPEND LIBS Boost::regex)
 else()
   set(boost_components)
   hunter_add_package(Boost)
 endif()
 find_package(Boost ${hunter_config} REQUIRED ${boost_components})
-list(APPEND package_deps Boost)
 
 hunter_add_package(thrift)
-find_package(thrift ${hunter_config} REQUIRED)
-if(HUNTER_ENABLED)
-  list(APPEND LIBS thrift::thrift_static)
-else()
-  list(APPEND LIBS ${THRIFT_LIBRARIES})
+find_package(thrift 0.9.2 ${hunter_config} REQUIRED)
+if(thrift_VERSION VERSION_GREATER "0.9.3")
```

Comment:
Maybe add `OR VERSION_LESS "0.9.2"`.

Reply:
That's handled by `find_package`. It doesn't have good support for saying "these two versions ONLY" hence the warning.

Reply:
Right. I guess my point is that `find_package` works differently with Hunter vs. non-Hunter.

Here's a brief overview. CMake offers two ways of finding packages (and maybe a third but IDK about it): modules and configs. Modules refer to the `Find<Package>` modules that you added here and that many people use in CMake builds. The problem is inherent redundancy of many projects. Why does everyone need to add these modules? So CMake offered another approach whereby the package declares its location during its CMake build, then other packages can easily locate it. The latter method is called the config approach.

Hunter relies SOLELY on the config approach. That means none of your `Find<Package>` files will be called at all. Relying on any code in there isn't going to work.

Reply:
Yeah, I realise that. That's why `${hunter_config}` substitutes the `CONFIG` option to `find_package`.

Have a look at the CMake package config files CMake generates from the Hunter-built packages. You'll see that they `add_library(blah SHARED IMPORT)` and setting `INTERFACE` properties like `INTERFACE_INCLUDE_DIRECTORIES` and `INTERFACE_INCLUDE_DIRECTORIES` on the resulting target. That's exactly what I'm doing in the `Find` packages too, to be consistent with CMake `CONFIG` mode. Look at the bundled CMake modules in `/usr/share/cmake/Modules` (or wherever yours are installed) e.g. `FindGTest.cmake`. You'll find that they *also* declare import targets and set properties on them for the library name, header paths, etc.

So it's not really to do with Hunter or non-Hunter. There are two things at play here:

* Is it using `CONFIG` mode (as Hunter requests), `MODULE` mode where it only uses `Find` packages, or the default where it tries a module first then falls back to config-mode?

* Does the resulting configuration created by a CONFIG file load or a Find module execution generate import-targets (new-style CMake), or just set variables (old-style CMake)?

What I've done is written the Find modules to produce the same import targets, etc, as what you get from loading the `CONFIG` module. So it all "Just Works". It's done how CMake's own bundled modules do it.

None of this is helped by the fairly erratic CMake documentation, but have a look at:

* https://cmake.org/cmake/help/v3.3/command/add_library.html#imported-libraries
* https://cmake.org/cmake/help/v3.3/prop_tgt/IMPORTED.html#prop_tgt:IMPORTED
* https://cmake.org/cmake/help/v3.3/manual/cmake-packages.7.html
* https://cmake.org/cmake/help/v3.0/prop_tgt/INTERFACE_INCLUDE_DIRECTORIES.html
* https://cmake.org/cmake/help/v3.0/prop_tgt/INTERFACE_LINK_LIBRARIES.html
* https://cmake.org/cmake/help/v3.0/prop_tgt/INTERFACE_COMPILE_DEFINITIONS.html

CMake takes care of the dependency chains, adding link libraries and include paths when you reference an import target, etc. It uses the properties set on the target for this. There's no need for a `LIBS` variable or any of that anymore.

This is also why the build-tree passed to headers in `target_include_directories` are `PRIVATE` and the install-path headers are `INTERFACE`. This tells CMake not to include the build-tree headers in the CMake config file it generates. And to add the install-target headers in the config file, but not use them when compiling the package its self, since it won't be installed yet.

CMake really needs a docs section that properly explains `PRIVATE`, `PUBLIC` and `INTERFACE`, explains `IMPORT` targets and how they're used in `Find` modules, etc. It's there, you just have to find all the scattered pieces from the docs and work out the whole picture from it. Dealing with this has been a massive pain, but hopefully it's nearly done.